### ISSUE TITLE
Cross-node updated data resulted in inconsistent data

### ISSUE THREAD

**Issue by osdba** [postgres-x2/postgres-x2#255]
### Environment：
- pgxc version: 1.2.1
- One gtm
- Three Datanodes:dn1,dn2,dn3
- Three Coordinate nodes: cd1,cd2,cd3
### Create test table

```
create table test02(id int primary key, x int);
insert into test02  select id as id1, id as id2 from (select generate_series(1,20) as id) t;
```
### Test

**Note**:  id = 6 row in dn1 node, id=7 row in dn2 node, then begin update two data cross node  cd1 and cd2. 

In cd1 node：

```
begin;
update test02 set x=16 where id=6;
```

in cd2：

```
begin;
update test02 set x=27 where id=7;
```

goto cd1 node：

```
update test02 set x=17 where id=7;
```

you can see aboved update statement blocked.

goto cd2 node：

```
update test02 set x=16 where id=6;
```

Strange things happened: above update not block. then you execute "commit", blocked statement in cd1 node not blocking.

goto cd1 node execute 'commit'.

then you query data in test02, you can see **two row data** that id=6:

```
postgres=# select * from test02;
 id | x
----+----
 ...
 ...
 18 | 18
 19 | 19
  6 | 16
  6 | 16
(21 rows)

postgres=# select * from test02 where x=16;
 id | x
----+----
  6 | 16
  6 | 16
 16 | 16
(3 rows)
```

**why can see two row(id=6), id is primary key?**

but if you query data using distributed key, you can see only on row:

```
postgres=# select * from test02 where id=6;
 id | x
----+----
  6 | 16
(1 row)
```

**Comment by Jasonysli** [comment]
I use the master code to do the same test, but I got two transactions blocked in step 2. The issue looks caused by uncorrect locator, which means the unblocked sql was sent to a wrong node, so the statement was not blocked.This kind of lock management was the same as PG original.

**Comment by koichi-szk** [comment]
Very interesting.   Do you know if we can reproduce the same lock
management problem in PG or do you have any reference to this issue in PG?
  We all depend upon PG's lock management for consistent update of
replicated tables.

Regards;

---

Koichi Suzuki

2015-07-01 16:54 GMT+09:00 Jasonysli notifications@github.com:

**Comment by Jasonysli** [comment]
What I mean is PGXC depends on original PG lock management to handle this situation, and it can not be reproduced  by using the latest code.

**Comment by koichi-szk** [comment]
What do you mean?   Do you mean that the latest PG fixed this problem?

---

Koichi Suzuki

2015-07-01 23:22 GMT+09:00 Jasonysli notifications@github.com:

**Comment by Jasonysli** [comment]
NO, I mean using latest code, the problem can't be reproduced and in postgres this situation is well handled.

**Comment by koichi-szk** [comment]
Okay, so at the latest X2 (XC) code, you cannot reproduce the problem.

Thank you;

---

Koichi Suzuki

2015-07-02 11:57 GMT+09:00 Jasonysli notifications@github.com:

**Comment by dierbua** [comment]
this bug is same as bug 239.

**Comment by dierbua** [comment]
please, give the coord,gtm and datanode logfile,thanks.

**Comment by koichi-szk** [comment]
This seems to be the same cause as we see in XL.    Pavan gave nice
analysis of the code and I'm discussing with him how to fix it.   It is not
simple but I believe we can fix it.

Thank you;

---

Koichi Suzuki
https://www.linkedin.com/in/koichidbms

2015-07-14 13:13 GMT+09:00 dierbua notifications@github.com:

**Comment by koichi-szk** [comment]
Please test with the latest REL1_2_STABLE branch.   I ran the conflict parallel update test from Konstantin and I did not observe any inconsistencies.

This looks duplicate to the issue #266.

**Comment by DBInsight** [comment]
we can use pgbench to test data inconsistency:
pgbench -i -h localhost -p 9122 -U XX bench_test ---init
pgbench -c 100 -T 300000 -n -h localhost -p 9122 -U XX bench_test > file.out  2>&1 --run

when running pgbench, use a another session to query data inconsistency:

select \* from pgbench_accounts a,(select aid,sum(delta) sumd from pgbench_history group by aid) h where h.sumd != a.abalance and h.aid=a.aid;

we may loop the above query when running pgbench to see if there is data inconsistency
